%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%


\documentclass[unnumsec,webpdf,namedate,modern,large]{oup-authoring-template}%

% Hyperlinks
\usepackage{hyperref}
\hypersetup{
    colorlinks=true,
    urlcolor=blue,
    citecolor=black,
    linkcolor=blue,
    }

\usepackage{graphicx}
\setcounter{figure}{0}
\graphicspath{{figures/}}

% New command to hyperlink figures
% See https://tex.stackexchange.com/questions/266434/extend-the-hyperref-link-to-figure-and-a-b-c
\newcommand*{\figref}[2][]{%
  \hyperref[{fig:#2}]{%
    Fig~\ref*{fig:#2}%
    \ifx\\#1\\%
    \else
      \,#1%
    \fi
  }%
}

% Reduce vertical spacing between sections
\usepackage{titlesec}
\titlespacing*{\section}{0pt}{0.2\baselineskip}{0.2\baselineskip}

%%% Bibliography %%%
\usepackage[style=authoryear,backend=biber,maxnames=1,uniquelist=false]{biblatex}
\addbibresource{references.bib}
                           
%% Bibiography style for articles %%%
\DeclareFieldFormat[article]{title}{#1} % Remove quotes from article titles
\DeclareFieldFormat[article]{volume}{\mkbibbold{#1}} % Make volume bold
\DeclareFieldFormat[article]{pages}{#1} % Remove "pp" abbreviation before the page number

% Customize bibliography according to Mini Oxford SCIMED style for Bioinformatics journal
\DeclareBibliographyDriver{article}{%
  \printnames{author}%%
  \newunit
  \printfield{title}.%
  \newunit
  \printfield{journaltitle}%
  \setunit*{\space}%
  \printfield{year}%
  \iffieldundef{volume}
    {\adddot} % If volume is undefined, end with a dot
    {%
      \addsemicolon\printfield{volume} % Add a semicolon, print volume
      \iffieldundef{pages}
        {\adddot} % If pages are undefined, end with a dot
         {\addcolon\printfield{pages}\adddot} % If pages are defined, print colon, pages, and end with a dot
     }
 }

%% Bibiography style for manuals %%%
\DeclareFieldFormat[manual]{volume}{\mkbibbold{#1}} % Make volume bold
\DeclareFieldFormat[manual]{pages}{#1} % Remove "pp" abbreviation before the page number

% For beautiful tables
\usepackage{booktabs}

% Remove white space above and below the tables
\usepackage{float}
\setlength{\floatsep}{10pt}
\setlength{\textfloatsep}{10pt}
\setlength{\intextsep}{10pt}

% Reduce blank space between sections
\usepackage[skip=1pt]{parskip}

\begin{document}

\journaltitle{Bioinformatics}

%\title{Orchestrating holo-omics with data containers}

%\title{Facilitating holo-omics with statistical programming}

%\title{Statistical programming framework for holo-omics}

%\title{Statistical workflows for holo-omics data integration}

\title{HoloFoodR: a statistical programming framework for holo-omics data integration workflows}

\firstpage{1}

\author[1,\dag]{Tuomas Borman\ORCID{0000-0002-8563-8884}}
\author[2,\dag]{Artur Sannikov\ORCID{0000-0001-7765-123X}}
\author[3]{Robert D. Finn\ORCID{0000-0001-8626-2148}}
\author[4]{Morten Tønsberg Limborg\ORCID{0000-0002-7718-6531}}
\author[5]{Alexander B. Rogers\ORCID{0000-0002-4283-6135}}
\author[6]{Varsha Kale\ORCID{0000-0002-3503-6383}}
\author[7]{Kati Hanhineva\ORCID{0000-0001-6834-7375}}
\author[8]{Leo Lahti\ORCID{0000-0001-5537-637X}}

\address[1,8]{\orgdiv{Department of Computing}, \orgname{University of Turku}, \postcode{20014} \state{Turku}, \country{Finland}}
\address[2,7]{\orgdiv{Department of Life Technologies}, \orgname{University of Turku}, \postcode{20014} \state{Turku}, \country{Finland}}
\address[3,5,6]{\orgdiv{Wellcome Genome Campus}, \orgname{EMBL-EBI}, \postcode{CB10 1SA} \state{Hinxton, Cambridgeshire}, \country{United Kingdom}}
\address[4]{\orgdiv{Center for Evolutionary Hologenomics, GLOBE Institute, Faculty of Health and Medical Sciences}, \orgname{University of Copenhagen}, \postcode{1353} \state{Copenhagen}, \country{Denmark}}
\corresp{\dag = Equal contribution.}

% Abstract
\abstract{
\textbf{Summary:} Holo-omics is an emerging research area that integrates multi-omic datasets from the host organism and its microbiome to study their interactions. Recently, curated and openly accessible holo-omic databases have been developed. The HoloFood database, for instance, provides nearly 10,000 holo-omic profiles for salmon and chicken under controlled treatments. However, bridging the gap between holo-omic data resources and algorithmic frameworks remains a challenge. Combining the latest advances in statistical programming with curated holo-omic data sets can facilitate the design of open and reproducible research workflows in the emerging field of holo-omics.
\\ % New line
\textbf{Availability and implementation:} HoloFoodR R/Bioconductor package and the source code are available under the open-source Artistic License 2.0 at the package homepage \url{https://github.com/EBI-Metagenomics/HoloFoodR}.
\\ % New line
\textbf{Contact:} microbiome@utu.fi
\\ % New line
\textbf{Supplementary information:} Supplementary data is available in the package vignette via the package homepage \url{https://ebi-metagenomics.github.io/HoloFoodR/articles/HoloFoodR.html}
}
%\textbf{Motivation:} Holo-omics is an emerging research area that integrates multi-omic data from the host organism and its microbiome to study their interactions. Recently, curated and openly accessible holo-omic databases have been developed. The HoloFood database, for instance, provides nearly 10,000 holo-omic profiles for salmon and chicken under controlled treatments. Despite the improvements in data availability, bridging the gap between holo-omic data resources and algorithmic frameworks remains a challenge. Method development in this area can facilitate the sharing of methods and best practices in the field of holo-omics.
%\\ % New line
%\textbf{Results:} We developed a data science framework to address this gap, supported by the new R/Bioconductor package, HoloFoodR. The package has been designed to streamline data input into a standardized format for downstream analysis within the Bioconductor ecosystem, which offers optimized solutions for individual omics as well as their integrative analysis. We demonstrate our methodology through a case study where we retrieve fatty acid data with associated metadata from the HoloFood database. It is  then integrated with with metagenomic data from the MGnify database, finally transforming the result into a dedicated multi-omic data {\it container}. This provides systematic means to jointly represent, manipulate, analyse rich combinations of omics and associated side information on the features and samples. We conclude by demonstrating the use of this data science strategy in exploratory analysis of host-microbiome interactions in a salmon model.
%\\
%\textbf{Availability and implementation:} HoloFoodR is freely available from \url{https://bioconductor.org/packages/release/bioc/html/HoloFoodR.html} under open-source Artistic License 2.0.}

\keywords{holo-omics, multiomics, metagenomics, metabolomics, data integration, bioconductor}

\maketitle

\section{Introduction}

The rapid advancement of omics technologies, including (meta)genomics and metabolomics, has been enhanced by breakthroughs in computational methods (\cite{moreno-indias_statistical_2021,marcos-zambrano_toolbox_2023,santamaria_bioinformatic_2024}).
These developments have enabled holo-omics, rapidly emerging field, which uses an integrative approach to comprehensively collect and analyze omic data from both a host organism and its associated microbiome, collectively referred to as the {\it holobiont} (\cite{nyholmHoloOmicsIntegratedHostMicrobiota2020,limborgAppliedHologenomicsFeasibility2018,odriozola_practical_2024}). The holo-omic approach has improved our understanding of complex biological systems, for example, in aquaculture (\cite{limborgAppliedHologenomicsFeasibility2018,atlantic_salmon}).

%This has enabled efficient simultaneous analyses of multiple types of omic data from a single sampleencompassing both host organisms and their diverse microbiomes (\cite{limborgAppliedHologenomicsFeasibility2018}).
%The integration of these diverse information sources has enhanced our understanding of complex diseases, metabolic pathways, and the impact of environmental factors on humans, animals, and plants (\cite{zhangProteogenomicCharacterizationHuman2014}).
%Progress in omics research focuses on discovering how different omics interact to impact phenotype.

%A recent review considered a holo-omic framework for fish (\cite{limborgAppliedHologenomicsFeasibility2018}) to identify potential microorganisms of importance in the production of beneficial fatty acids (e.g. eicosapentaenoic and docosahexaenoic acids) in fish muscles. This review, together with empirical case studies (\cite{atlantic_salmon}), suggested that aquaculture could be further optimized by combining holo-omic, dietary and environmental data.

Holo-omic research and the collaborative development of computational methods could benefit from curated, open access data resources.
%With the increase of data volume, expenses and time required for data collection rise.
The limited availability of comprehensive multi-omic data resources can impede research progress, constraining computational method development. Open data portals with an accessible Application Programming Interface (API) are crucial in overcoming these challenges. They facilitate research by providing access to data resources and supporting the collaborative development and benchmarking of new data science methods to extract insights from large, curated datasets (\cite{pasolli_2017}).

Data can be limited in value unless it adheres to the FAIR principles (Findability, Accessibility, Interoperability, Reuse) (\cite{wilkinsonFAIRGuidingPrinciples2016}). API interfacing and data wrangling require specialized skills, leading to non-transferable, error-prone workflows, not easily reusable by the wider scientific community. Therefore, standardized open-source workflows are needed to search, retrieve and convert data into a suitable format for downstream analyses.
To narrow the gap between upstream data retrieval and downstream analysis, we developed the HoloFoodR package, which facilitates seamless programmatic linking of the holo-omic data via the HoloFood data portal API and analysis methods from the R/Bioconductor  (\cite{gentleman_bioconductor_2004,callahan_bioconductor_2016}).
%Towards this objective, we are leveraging the the R/Bioconductor methods ecosystem --- an open-source project for developing computational biology tools.

\section{Materials and methods}

The HoloFood data portal is an open access portal of curated holo-omic data and analyses, developed by the international HoloFood consortium (\cite{rogers_holofood_2025}) and hosted by the European Bioinformatics Institute (EMBL-EBI). It centralizes access to heterogeneous data resources including the European Nucleotide Archive for sequence data, MGnify for metagenomic data, and MetaboLights for metabolomic data, and tracks their interrelations via a web portal and free API, covering nearly 10,000 samples from over 2,000 individual chickens and salmon. Biomolecular and physiological measurements were collected at the level of each individual specimen in the project to explore the effects of novel sustainable feeds on physiological processes in farmed animals. In addition, the metadata of each sample is stored in the EMBL-EBI BioSamples service (\cite{courtot_2018}).

We developed a conceptual framework to support best practices of statistical programming in holo-omic research, and implemented the work as the HoloFoodR R software library. The package relies on specialized data containers that have been designed to organize the multi-omic data in a structured format. Once the data has been imported to these formats, users can leverage state-of-the-art statistical programming methods for data processing and analysis.

\vspace{-1em}
\subsection{Integrative data containers}

Data containers provide structured, standardized storage for data, proving useful in life science informatics where complex, hierarchical and multi-source data collections are common to describe the studied phenomenon (\cite{drnevich_2024}). Custom data containers simplify handling diverse omics data, supporting the design of modular and reproducible workflows.

Two data structures are central to our approach, encompassing the needs to organize data for a single omics type, and subsequently link these across multiple omics.
First, the TreeSummarizedExperiment (TreeSE) data container (\cite{huang_treesummarizedexperiment_2021}) standardizes single-omic data, such as metagenome or metabolome derived data sets. It stores experimental data, along with feature and sample metadata, in a structured format. Additionally, it supports the integration of hierarchical information on the features and samples, such as phylogenetic trees. The second container, MultiAssayExperiment (MAE) (\cite{ramos_software_2017}) acts as a logical complement by integrating multiple, heterogeneous omics datasets, seamlessly linking samples across single-omic datasets. Together, these data containers form the basis for an ecosystem of interoperable methods developed by the R/Bioconductor community.

\vspace{-1em}
\subsection{HoloFoodR workflow}

The HoloFoodR package structures data in these single and multi-omic formats, enabling efficient data importing from holo-omic databases for streamlined analysis and integration. The HoloFoodR workflow is summarized in \figref{workflow}. The package uses four functions to query and retrieve data from the HoloFood and MetaboLights databases (Table \ref{tab:holofoodr_functions}). A key feature is its ability to organize versatile data combinations into the MAE data container. Thus, it transforms complex data from the HoloFood API into a standardized structure for downstream statistical analyses in the Bioconductor data science environment  (\cite{ramos_software_2017}).

Furthermore, HoloFoodR leverages the MetaboLights database to retrieve non-targeted metabolomic datasets indicated by project identifiers in the HoloFood data. The data can then be integrated with complementary metagenomic data resources that are available via the MGnify database using the MGnifyR package (\cite{mgnifyr}).

\vspace{-1em}

\begin{table}[H]
\begin{center}
\begin{minipage}{\linewidth}
\caption{HoloFoodR functions that facilitate data search and retrieval}\label{tab:holofoodr_functions}
    \begin{tabular}{l p{0.6\linewidth}}
        \toprule
        Function & Description \\ 
        \midrule
        doQuery() & Query HoloFood database \\
        getData() & Retrieve sample data in list or data.frame format \\
        getResult()  & Retrieve sample data in MultiAssayExperiment format \\
        getMetaboLights() & Retrieve non-targeted metabolomic data from MetaboLights \\
        \bottomrule
    \end{tabular}
\end{minipage}
\end{center}
\end{table}

\vspace{-1em}

\begin{figure*}[t]
	\centering
    % \vspace{20pt}%
    \includegraphics[width=1\textwidth]{holofoodr_workflow_9.png}
    \label{fig:workflow}
    \caption{Data science workflow for holo-omic analysis. Programmatic access to open omic databases is facilitated by the dedicated software that feeds the data to R computational environment. The versatile R/Bioconductor is utilized to visualize microbial genera abundances and to analyze  time effect on omega-3 fatty acid concentrations in salmon muscles. The illustration and logo of the multi-assay data container have been taken from the MultiAssayExperiment vignette (Artistic License 2.0).}
	\label{fig:workflow}
\end{figure*}

\vspace{-1em}
\subsection{Bioconductor methods ecosystem}

When data from the HoloFood database is retrieved via the HoloFoodR package, users gain direct access to single- and multi-omic downstream analysis methods available in Bioconductor. It currently offers 2,300 packages, developed by a diverse community of individual contributors for a wide range of fields (see e.g. \cite{gentleman_bioconductor_2004,callahan_bioconductor_2016,amezquita_orchestrating_2020,drnevich_2024}). Of particular relevance for studying host-microbiome interactions, Bioconductor includes the specifically designed \texttt{mia} framework centering around these containers and supports a range of analysis and visualization methods in microbiome research (\cite{mia, miaViz, oma}). Similarly, the metabolomic data can be pre-processed, analyzed and visualized with the \texttt{notame} package (\cite{klavus_2020}).

\section{Results}

In this section, we present a practical use case demonstrating the application of HoloFoodR within a holo-omics workflow. The workflow includes the following steps:

\begin{enumerate}
    \item Fetch and integrate data from the HoloFood and MGnify databases.
    \item Filter, clean, and transform data for analysis.
    \item Explore and summarize the data.
    \item Test associations between fatty acids, time, and treatment.
    \item Test associations between microbiome composition, time, and treatment.
    \item Characterize the joint variation between the parallel omics measurements.
\end{enumerate}

The full workflow including figures and data refinement is available in the package vignette at \url{https://ebi-metagenomics.github.io/HoloFoodR/articles}.

\vspace{-1em}
\subsection{Data import and wrangling}

The workflow begins by querying the HoloFood database to retrieve available animals. We chose to gather all available salmon entries, collecting metadata on these animals to explore the available data types and samples for each. These samples were then compiled into MAE data container for further analysis. For metagenomic data retrieval, we leveraged the capabilities of the MGnifyR package to first identify the corresponding sample IDs from the MGnify database and then fetch the metagenomic data as a MAE object. The consistent data format supports  seamless merging of the datasets from HoloFood and MGnify. The resulting data container carries various types of omics with associated metadata, and provides the basis for downstream analyses.

By employing the capabilities of the MAE object, we efficiently wrangled the data to focus on salmon from Trial A in the HoloFood database, to study the impact of fermented seaweed as a feed additive. Samples were obtained from animals euthanized during the sampling process, with animals assigned to be sampled either at the beginning or at the end of the trial. Fatty acid concentrations were analyzed from muscle tissue, while metagenomics samples were drawn from the intestine.

The \texttt{mia} framework (\cite{mia, miaViz}) provides a user-friendly R interface designed for the downstream analysis, with a particular emphasis on microbiome data. We filtered and agglomerated the data to concentrate on a specific group of microbes and fatty acids. Additionally, various data transformations were applied to prepare the dataset for further analysis. Briefly, we have transformed metagenomic data with centered log ratio method, and fatty acid data with logarithmic transformation. The full details are provided in the package vignette.

\vspace{-1em}
\subsection{Downstream analysis}

After data retrieval and basic exploration, we focused on the effects of the seaweed treatment and aging. We did not observe any impact of treatment on the fatty acid composition, which aligns with the findings of a recent HoloFood study (\cite{rasmussen_holo-omics_2025}). However, we did note temporal effects. As the salmon grew, the concentration increased for certain fatty acids, such as docosahexaenoic, eicosapentaenoic, linoleic, oleic, palmitic, and stearic  (see \figref{workflow}, data analysis panel, which highlights the first two fatty acids).

Microbial community analysis indicated a dominance of {\it Mycoplasma}, consistent with previous studies ({\cite{zarkasiPyrosequencingbasedCharacterizationGastrointestinal2014,bozzi_salmon_2021}}, see \figref{workflow}, data analysis panel). We also observed an increase in microbial diversity (Shannon index) with time, confirmed by Principal Coordinate Analysis (Bray-Curtis dissimilarity).

To explore the co-abundance of fatty acids and microbial genera, we performed multi-omic factor analysis (\cite{argelaguet_mofa_2020}), revealing that {\it Cetobacterium}, {\it Vibrio}, {\it Aliivibrio}, and {\it Photobacterium} covaried with overall fatty acid levels, while {\it Mycoplasma} showed no such correlation.

\section{Discussion and conclusions}
% Context / background / reflection on objectives

The integration of multiple omics layers can offer holistic insights into complex living systems. Holo-omics is the new field that emphasizes the interactions between the hosts and their microbiomes. However, such analyses rely on access to curated data collections, computational environments and algorithmic tools. The evolving field of holo-omics can benefit greatly from the collaborative methods development and sharing, thus avoiding the duplication of work (\cite{open_microbiome, lahti_open_2018}).

To fill the gap between holo-omic data and downstream analysis methods, we developed the HoloFoodR package. It provides programmatic access to the HoloFood data portal and links it with complementary data sources analysis methods from the extensive Bioconductor ecosystem. Our work demonstrates a comprehensive open data science strategy for holo-omics research.

The HoloFoodR software library simplifies data retrieval by minimizing external data wrangling adhering to standardized data structures that support efficient use of the multi-omic analyses methods. This standardization replaces lengthy \textit{ad-hoc} code with concise, and tested open-source solutions. We illustrate the approach with an end-to-end workflow that covers typical steps of a holo-omic data analysis and emphasizes the use of multi-assay data containers to enhance access to methods in the Bioconductor ecosystem.

% Use in teaching
In addition to method development and analysis, we see potential in using HoloFood data in teaching multi-omics techniques. HoloFoodR offers simplified access to real-world omic data, enabling more proficient users "learn by doing”. The advantage of such data over “toy” data lies in its increased complexity, which provides an excellent foundation for instructors to teach students advanced data cleaning and analysis techniques (\cite{drnevich_2024}).

The HoloFood data portal currently provides the interconnectivity between the samples and omics datasets, based on BioSamples identifiers. This linkage provides a template for future multiomics datasets, and as the number of datasets increases, the HoloFoodR package could be generalised so that the concepts can be readily applied to other datasets.

% Limitations, implementation issues
Despite the benefits, there are also limitations. First, raw spectral metabolite data from the MetaboLights database requires extensive preprocessing, often with external tools (\cite{klavus_2020}). Whereas the proposed custom data structures can readily support downstream analyses, their construction and use require sufficient R programming skills.

% Concluding sentences
Thus, the methods and open data science strategy that we have suggested can serve as a template for conducting multi-omic analyses. The HoloFoodR package can be adapted for other data resources relevant to holo-omic research, where the growing Bioconductor ecosystem offers an expanding compilation of data analytical tools. 

\section{Acknowledgements}

We are grateful to HoloFood consortium (\url{https://www.holofood.eu}) for providing the curated database and API.

% \section{Supplementary data}

\section{Competing interests}
No competing interest is declared.

\section{Funding}
This work was supported by the European Commission in the framework of the Horizon2020 Project FindingPheno [GA  952914] and HoloFood [GA 817729] . A.S. and K.H. were  supported by Jane and Aatos Erkko Foundation and the Research Council of Finland [grant numbers 321716, 334814]. M.T.L. was supported by the Danish National Research Foundation [grant DNRF143].  L.L. was supported by Research Council of Finland [grant number 330887].

\printbibliography

\end{document}